src/EditorFeatures/CSharpTest/MisplacedUsings/MisplacedUsingsCodeFixProviderTests.cs

```diff
@@ -0,0 +1,1295 @@
+
+// Leading Comment
+
+using System;
+using System.Threading;
+
+namespace TestNamespace
+{
+}
+";
+
+            var fixedTestCode = @"// File Header
+
+namespace TestNamespace
+{
+    // Leading Comment
+
+    using System;
+    using System.Threading;
+}
+";
+
+            DiagnosticResult[] expectedResults =
+            {
+                Diagnostic(MisplacedUsingsDiagnosticAnalyzer._insideDescriptor).WithLocation(5, 1),
+                Diagnostic(MisplacedUsingsDiagnosticAnalyzer._insideDescriptor).WithLocation(6, 1),
+            };
+
+            return VerifyCodeFixAsync(testCode, s_insidePreferenceOption, expectedResults, fixedTestCode, placeSystemNamespaceFirst: true, separateImportDirectiveGroups: false);
+        }
+
+        /// <summary>
+        /// Verifies that a code fix will be offered for MisplacedUsing diagnostics when a single namespace is present.
+        /// </summary>
+        [Fact]
+        public Task WhenInsidePreferred_UsingsInCompilationUnitWithSingleNamespace_UsingsMoved()
+        {
+            var testCode = @"using System;
+
+namespace TestNamespace1
+{
+    public class TestClass1
+    {
+    }
+}
+";
+            var fixedTestCode = @"namespace TestNamespace1
+{
+    using System;
+
+    public class TestClass1
+    {
+    }
+}
+";
+
+            DiagnosticResult[] expected =
+            {
+                Diagnostic(MisplacedUsingsDiagnosticAnalyzer._insideDescriptor).WithLocation(1, 1)
+            };
+            return VerifyCodeFixAsync(testCode, s_insidePreferenceOption, expected, fixedTestCode, placeSystemNamespaceFirst: true, separateImportDirectiveGroups: false);
+        }
+
+        /// <summary>
+        /// Verifies that a code fix will not be offered for MisplacedUsing diagnostics when multiple namespaces are present.
+        /// </summary>
+        [Fact]
+        public Task WhenInsidePreferred_UsingsInCompilationUnitWithMultipleNamespaces_NoCodeFixOffered()
+        {
+            var testCode = @"using System;
+
+namespace TestNamespace1
+{
+    public class TestClass1
+    {
+    }
+}
+
+namespace TestNamespace2
+{
+}
+";
+            DiagnosticResult[] expected =
+            {
+                Diagnostic(MisplacedUsingsDiagnosticAnalyzer._insideDescriptor).WithLocation(1, 1)
+            };
+            return VerifyCodeFixAsync(testCode, s_insidePreferenceOption, expected, fixedSource: testCode, remaining: expected, placeSystemNamespaceFirst: true, separateImportDirectiveGroups: false);
+        }
+
+        /// <summary>
+        /// Verifies that the code fix will properly move pragmas.
+        /// </summary>
+        [Fact]
+        public Task WhenInsidePreferred_UsingsInCompilationUnitWithPragma_PragmaMoved()
+        {
+            var testCode = @"#pragma warning disable 1573 // Comment
+using System;
+using System.Threading;
+
+namespace TestNamespace
+{
+}
+";
+
+            var fixedTestCode = @"namespace TestNamespace
+{
+#pragma warning disable 1573 // Comment
```

Comment:
why does the pragma move with the usings?

Reply:
:arrow_right: It could be suppressing a diagnostic related to the using directive itself

Reply:
wouldn't that still happen even if it didn't move?  what if its suppressing something related to the namespace it's moving into?